# Police  Dataset
Dataset collected from the Police check-post

In [1]:
import pandas as pd

In [59]:
data=pd.read_csv(r"C:\Users\Shashi Shekhar\Desktop\python\Dataset for resume projects\Police Dataset.csv")


In [60]:
data.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

In [61]:
data.shape#getting values of rows and columns

(65535, 15)

### Q1. remove the columns which has only missing values

In [62]:
data.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

As above we can analyse that only one column "country_name" have only missing values

In [63]:
data.drop(columns=["country_name"], inplace=True)

In [64]:
data.shape #getting result by removing column having only missing values

(65535, 14)

### Q2. for speeding were men/women stopped more often?

In [65]:
data.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

In [66]:
data[data["violation"]=="Speeding"].driver_gender.value_counts() # df[df["col1"]==""].col2.value_counts(), getting result

driver_gender
M    25517
F    11686
Name: count, dtype: int64

men were stopped more often for speeding as from above analysis

### Q3. Does gender affect who get searched during stop

In [67]:
data.groupby("driver_gender").search_conducted.sum()#using "df.groupby("col1").col2.sum()"

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

In [68]:
data.search_conducted.value_counts() #using "df.col1.value_counts()" to get the count of each value of col1

search_conducted
False    63056
True      2479
Name: count, dtype: int64

from this record we analyse that Yes, gender affect when get seached during stop as men(2113)>> women(366) as True(2113+366=2479)

### Q4. what is the mean stop duration?

In [69]:
data.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

In [70]:
#que based on mapping + data type casting
data.stop_duration.value_counts()#using "df.col1.value_counts()" to get the count of each value of col1

stop_duration
0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: count, dtype: int64

In [71]:
#use mapping command as "df["col1"]=df["col1"].map({old1:new1, old2:new2,})"
data["stop_duration"]=data["stop_duration"].map({"0-15 Min":7.5, "16-30 Min":24, "30+ Min":45})#uses mean value for new

In [72]:
data.stop_duration.nunique()

3

In [73]:
data.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested  stop_duration  drugs_related_stop  
0       False            7.5               False  
1       False            7.5               False  
2       False            7.5               False  
3        True           24.0               False  
4       False            7.5               False

In [76]:
round(data["stop_duration"].mean(),2)#getting answer

12.19

### Q5. Compare the age distribution of each violation

In [79]:
data.groupby("violation").driver_age.value_counts()#getting the age count for each violation

violation  driver_age
Equipment  23.0          385
           21.0          374
           22.0          370
           20.0          348
           24.0          329
                        ... 
Speeding   84.0            3
           81.0            3
           88.0            2
           83.0            1
           15.0            1
Name: count, Length: 333, dtype: int64

In [81]:
data.groupby("violation").driver_age.describe()#getting the complete comparision of driver age for each violation

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0